File: src/importer/email-normalizer.ts
import path from 'path';
import fs from 'fs';
import { Email, EmailAddress } from '../models/email';
import { Attachment } from '../models/attachment';

export interface NormalizedEmail {
  email: Email;
  attachments: RawAttachment[];
}

export interface RawAttachment {
  filename: string;
  mimeType?: string;
  data: Buffer;
}

export function parseAddressList(raw: string | null | undefined): EmailAddress[] {
  if (!raw) return [];
  // Handle semicolon or comma separated lists like "Name <email>, Name2 <email2>"
  return raw
    .split(/[;,]/)
    .map((part) => part.trim())
    .filter(Boolean)
    .map(parseAddress);
}

export function parseAddress(raw: string): EmailAddress {
  const match = raw.match(/^(.+)\s*<([^>]+)>$/);
  if (match) {
    return { name: match[1].trim(), email: match[2].trim() };
  }
  return { email: raw.trim() };
}

export function buildAttachment(
  emailId: number,
  raw: RawAttachment,
  attachmentDir: string
): Attachment {
  if (!fs.existsSync(attachmentDir)) {
    fs.mkdirSync(attachmentDir, { recursive: true });
  }

  const safeName = raw.filename.replace(/[/\\:*?"<>|]/g, '_');
  let storagePath = path.join(attachmentDir, safeName);

  // Avoid overwriting existing files
  let counter = 1;
  while (fs.existsSync(storagePath)) {
    const ext = path.extname(safeName);
    const base = path.basename(safeName, ext);
    storagePath = path.join(attachmentDir, `${base}_${counter}${ext}`);
    counter++;
  }

  fs.writeFileSync(storagePath, raw.data);

  return {
    emailId,
    filename: raw.filename,
    mimeType: raw.mimeType,
    size: raw.data.length,
    storagePath,
  };
}

File: src/MailSearch/Importer/EmailNormalizer.cs
using System.Text.RegularExpressions;
using MailSearch.Models;

namespace MailSearch.Importer;

public static partial class EmailNormalizer
{
    /// <summary>
    /// Parses a comma- or semicolon-separated list of RFC 5322-style address strings.
    /// </summary>
    public static List<EmailAddress> ParseAddressList(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw)) return [];

        return raw
            .Split([';', ','])
            .Select(p => p.Trim())
            .Where(p => p.Length > 0)
            .Select(ParseAddress)
            .ToList();
    }

    /// <summary>
    /// Parses a single RFC 5322-style address, e.g. "Alice Smith &lt;alice@example.com&gt;" or "alice@example.com".
    /// </summary>
    public static EmailAddress ParseAddress(string raw)
    {
        raw = raw.Trim();
        var match = NamedEmailRegex().Match(raw);
        if (match.Success)
        {
            return new EmailAddress
            {
                Name = match.Groups["name"].Value.Trim(),
                Email = match.Groups["email"].Value.Trim(),
            };
        }
        return new EmailAddress { Email = raw };
    }

    /// <summary>
    /// Saves a raw attachment to disk (avoiding filename collisions) and returns an <see cref="Attachment"/> record.
    /// </summary>
    public static Attachment BuildAttachment(int emailId, string filename, string? mimeType, byte[] data, string attachmentDir)
    {
        if (!Directory.Exists(attachmentDir))
            Directory.CreateDirectory(attachmentDir);

        // Strip characters that are unsafe in file names
        var safeName = InvalidFileNameCharsRegex().Replace(filename, "_");
        var storagePath = Path.Combine(attachmentDir, safeName);

        // Avoid overwriting existing files
        if (File.Exists(storagePath))
        {
            var ext = Path.GetExtension(safeName);
            var baseName = Path.GetFileNameWithoutExtension(safeName);
            int counter = 1;
            do
            {
                storagePath = Path.Combine(attachmentDir, $"{baseName}_{counter}{ext}");
                counter++;
            }
            while (File.Exists(storagePath));
        }

        File.WriteAllBytes(storagePath, data);

        return new Attachment
        {
            EmailId = emailId,
            Filename = filename,
            MimeType = mimeType,
            Size = data.Length,
            StoragePath = storagePath,
        };
    }

    [GeneratedRegex(@"^(?<name>.+)\s*<(?<email>[^>]+)>$")]
    private static partial Regex NamedEmailRegex();

    [GeneratedRegex(@"[/\\:*?""<>|]")]
    private static partial Regex InvalidFileNameCharsRegex();
}
